src/python/pants/backend/jvm/tasks/coverage/scoverage.py

```diff
@@ -0,0 +1,201 @@
+# coding=utf-8
+# Copyright 2014 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import functools
+import os
+import re
+
+from pants.backend.jvm.subsystems.jvm_tool_mixin import JvmToolMixin
+from pants.java.jar.jar_dependency import JarDependency
+from pants.backend.jvm.tasks.coverage.engine import CoverageEngine
+from pants.base.exceptions import TaskError
+from pants.subsystem.subsystem import Subsystem
+from pants.util.dirutil import relativize_paths, safe_mkdir, safe_mkdir_for, safe_walk, touch
+
+
+
+class Scoverage(CoverageEngine):
+  """Class to run coverage tests with scoverage"""
+
+
+  class Factory(Subsystem, JvmToolMixin):
+
+    # Cannot have the same scope as ScoveragePlatform, i.e they
+    # both cannot share the scope `scoverage`.
+    options_scope = 'scoverage-runtime'
+
+    @classmethod
+    def register_options(cls, register):
+      super(Scoverage.Factory, cls).register_options(register)
+
+
+      def scoverage_jar(name, **kwargs):
+        return JarDependency(org='com.twitter.scoverage', name=name, rev='1.0.1-twitter', **kwargs)
+
+      def slf4j_jar(name):
+        return JarDependency(org='org.slf4j', name=name, rev='1.7.5')
+
+      def scopt_jar():
+        return JarDependency(org='com.github.scopt', name='scopt_2.12', rev='3.7.0')
+
+      def commons_jar():
+        return JarDependency(org='commons-io', name='commons-io', rev='2.5')
+
+      def scoverage_report_jar(**kwargs):
+        return JarDependency(org='org.pantsbuild', name='scoverage-report-generator_2.12',
+          rev='0.0.1', **kwargs)
+
+
+      # We need to inject report generator at runtime.
+      cls.register_jvm_tool(register,
+        'scoverage-report',
+        classpath=[
+        scoverage_report_jar(),
+        commons_jar(),
+        scopt_jar(),
+        slf4j_jar('slf4j-simple'), slf4j_jar('slf4j-api'),
+        scoverage_jar('scalac-scoverage-plugin_2.12')
+        ]
+      )
+
+      register('--target-filters', type=list, default=[],
```

Comment:
This feels redundant with the ScoveragePlatform.blacklist option. Do we need both?

Reply:
This target filtering is for report generation (similar to what jacoco does), i.e helps in generating reports only for the specified targets

Reply:
this option helps in preventing the need to compile again if the report is needed only for a handful of targets, i.e, say I compiled 500 targets and got the report generated. But now I need a report for all targets containing the class "merlin"(~ 10 targets), then I can do that without needing to compile everything again.

Reply:
Hm. But in reality, are people going to know that? From a user's perspective, having just one option would likely be easier to understand.

Reply:
Agreed. People just need to use this `target-filter` option most of the time. In case they encounter this error of JVM code size limits, they can overcome that with the help of 1blacklist` option as well. More like a Troubleshooting mechanism. I'll make sure to document this.